doc/pbh/pbh-design.md

```diff
@@ -0,0 +1,770 @@
+              / [ *3( h16 ":" ) h16 ] "::"    h16 ":"   ls32
+              / [ *4( h16 ":" ) h16 ] "::"              ls32
+              / [ *5( h16 ":" ) h16 ] "::"              h16
+              / [ *6( h16 ":" ) h16 ] "::"
+```
+
+**Note:** field _ipv4_mask_/_ipv6_mask_ is only valid when _hash_field_ equals _INNER_DST/SRC_IPV4/IPV6_
+
+### 2.4.2 Configuration sample
+
+**Inner 5-tuple hashing:**
+```json
+{
+    "PBH_HASH": {
+        "inner_ip_proto": {
+            "hash_field": "INNER_IP_PROTOCOL",
+            "sequence_id": "1"
+        },
+        "inner_l4_dst_port": {
+            "hash_field": "INNER_L4_DST_PORT",
+            "sequence_id": "2"
+        },
+        "inner_l4_src_port": {
+            "hash_field": "INNER_L4_SRC_PORT",
+            "sequence_id": "2"
+        },
+        "inner_dst_ipv4": {
+            "hash_field": "INNER_DST_IPV4",
+            "ipv4_mask": "255.0.0.0",
+            "sequence_id": "3"
+        },
+        "inner_src_ipv4": {
+            "hash_field": "INNER_SRC_IPV4",
+            "ipv4_mask": "0.0.0.255",
+            "sequence_id": "3"
+        },
+        "inner_dst_ipv6": {
+            "hash_field": "INNER_DST_IPV6",
+            "ipv6_mask": "FFFF::",
+            "sequence_id": "4"
+        },
+        "inner_src_ipv6": {
+            "hash_field": "INNER_SRC_IPV6",
+            "ipv6_mask": "::FFFF",
+            "sequence_id": "4"
+        }
+    },
+    "PBH_RULE": {
+        "pbh_table|nvgre": {
+            "priority": "1",
+            "gre_key": "0x2500/0xffffff00",
+            "inner_ether_type": "0x86dd/0xffff",
+            "hash_list": [
+                "inner_ip_proto",
+                "inner_l4_dst_port",
+                "inner_l4_src_port",
+                "inner_dst_ipv6",
+                "inner_src_ipv6"
+            ],
+            "packet_action": "SET_ECMP_HASH"
+        },
+        "pbh_table|vxlan": {
+            "priority": "2",
+            "ip_protocol": "0x11/0xff",
+            "l4_dst_port": "0x12b5/0xffff",
+            "inner_ether_type": "0x0800/0xffff",
+            "hash_list": [
+                "inner_ip_proto",
+                "inner_l4_dst_port",
+                "inner_l4_src_port",
+                "inner_dst_ipv4",
+                "inner_src_ipv4"
+            ],
+            "packet_action": "SET_LAG_HASH"
+        }
+    },
+    "PBH_TABLE": {
+        "pbh_table": {
+            "port_list": [
+                "Ethernet0",
+                "Ethernet4"
+            ],
+            "lag_list": [
+                "PortChannel0001",
+                "PortChannel0002"
+            ],
+            "description": "NVGRE and VxLAN"
+        }
+    }
+}
+```
+
+## 2.5 Flows
+
+### 2.5.1 PBH add
+
+![PBH add flow](images/pbh_add_flow.svg "Figure 3: PBH add flow")
+
+###### Figure 3: PBH add flow
+
+### 2.5.2 PBH remove
+
+![PBH remove flow](images/pbh_remove_flow.svg "Figure 4: PBH remove flow")
```

Comment:
I see the workflow, there is no dependency which is good . However, shouldn't it be delete Rules before the Pbh table get deleted? How does it work if table get deleted and rules failed to deleted. 
How do we handling the error scenarios?

Reply:
@madhupalu the implementation assumes strict dependencies for all PBH objects: there is no way to remove PBH table if it's still being referenced by any PBH rule